### DISCUSSION TITLE
[Mini.clue] Get mappings from vim script plugin

### DISCUSSION THREAD

**Discussion by bassamsdata** [nvim-mini/mini.nvim#493] [Q&A] [answered]
I'm using vim script plugin which is for R language [NVIM-R](https://github.com/jalvesaq/Nvim-R) and it has around 50 or so mappings that all start with localleader `\`.

`Mini.clue` gets the mappings but doesn't get the descriptions (probably because it's vim script that has no `desc`) like below
<img width="434" alt="Screen Shot 2023-09-26 at 3 37 30 PM" src="https://github.com/echasnovski/mini.nvim/assets/105807570/e2154ab5-35aa-4836-a128-efd0da95e8ed">

I checked the source code of the mapping in `NVIM-R` plugin and it was like so:
```vim
   function RCreateCommentMaps()
        RCreateMapsWrapper("ni", "RToggleComment", "xx", ':call RComment("normal")')
        RCreateMapsWrapper("v", "RToggleComment", "xx", ':call RComment("selection")')
        RCreateMapsWrapper("ni", "RSimpleComment", "xc", ':call RSimpleCommentLine("normal", "c")')
        RCreateMapsWrapper("v", "RSimpleComment", "xc", ':call RSimpleCommentLine("selection", "c")')
        RCreateMapsWrapper("ni", "RSimpleUnComment", "xu", ':call RSimpleCommentLine("normal", "u")')
        RCreateMapsWrapper("v", "RSimpleUnComment", "xu", ':call RSimpleCommentLine("selection", "u")')
        RCreateMapsWrapper("ni", "RRightComment", ";", ':call MovePosRCodeComment("normal")')
        RCreateMapsWrapper("v", "RRightComment", ";", ':call MovePosRCodeComment("selection")')
      end
```

so the keymapping function is like that:
```vim
function RCreateMaps(type, plug, combo, target)
```

Because I don't want to write desc for 50 mappings, I'm okay with getting the `plug` part as `desc` so I wrote a wrapper function to get it but didn't work unfortunately. 

is there a way to do that with `mini.clue`, if not, I'll write description for the the keymapping but I'm not sure how to do that without affecting my config. I wan thinking of [mapx.nvim](https://github.com/b0o/mapx.nvim) plugin.
```lua
require('mapx').nnoremap('<localleader>r', ':RunRCommand<CR>', {desc = 'Run R command'})
```
any help would be really appreciated

**Comment by echasnovski** [comment]
'mini.clue' indeed uses mapping description to infer what information to shoe. The intentional decision is to [make mapping description take priority over manually defined clues in `config.clues`](https://github.com/echasnovski/mini.nvim/blob/1fdbb864e2015eb6f501394d593630f825154385/doc/mini-clue.txt#L79-L82). The suggested solution is to use [`set_mapping_desc()`](https://github.com/echasnovski/mini.nvim/blob/1fdbb864e2015eb6f501394d593630f825154385/doc/mini-clue.txt#L622) to update mapping description.

**Reply by bassamsdata** [reply]
thank you for your response and help. Unfortunately, whatever I do it just does't work. If you can help identify If I'm doing something wrong.
this first solution didn't work
```lua
-- Helper function to check if the R language server is attached to the current buffer
local function is_r_language_server_attached()
  local bufnr = vim.api.nvim_get_current_buf()
  local clients = vim.lsp.buf_get_active_clients(bufnr)
  for _, client in pairs(clients) do
    if client.name == "r_language_server" then
      return true
    end
  end
  return false
end

local miniclue = require("mini.clue")
if vim.bo.filetype == 'r' and is_r_language_server_attached() then
    miniclue.set_mapping_desc('n', '\\fa', 'My description')
end

```
and I did this:
```lua
-- Add an autocmd for the FileType event
vim.api.nvim_exec([[
  augroup MyLuaMaps
    autocmd!
    autocmd FileType r lua require('mini.clue').set_mapping_desc('n', '\\fa', 'My description')
  augroup END
]], false)

```
and this 
```lua
vim.api.nvim_create_autocmd({ "BufEnter" }, {
  pattern = { "*.r" },
  callback = function()
    local bufnr = vim.api.nvim_get_current_buf()

    local clients = vim.lsp.buf_get_clients(bufnr)
    for _, client in pairs(clients) do
      if client.name == 'r_language_server' then

        require('mini.clue').set_mapping_desc('v', '\\fa', 'Your description')
      end
    end
  end,
})

```
and this one 🫣
```lua
local nvim_lsp = require('lspconfig')
nvim_lsp.r_language_server.setup {
  on_attach = function(client, bufnr)
    -- Your code to create the key mapping goes here

    -- Then set the description for the mapping
    require('mini.clue').set_mapping_desc('v', '\\fa', 'Your description')
  end,
}
```
and this one as well inside the setup function after
` local miniclue = require("mini.clue")`
```lua 
        function()
          if vim.bo.filetype == "r" then
            -- If both conditions are met, set the description for the mapping
            miniclue.set_mapping_desc('v', '\\fa', 'Your description')
          end
        end,
```
**I probably leave it as it is but I need to know, am I doing something wrong.**

Thank you

**Reply by echasnovski** [reply, marked as answer]
That is a lot of dedicated effort from you, my fellow R user :)

I am not sure what might be wrong with your solutions. My first guess would be that they get executed *before* filetype mappings from 'Nvim-R' are created. Although that would result in an error from `set_mapping_desc()`... Not sure at all.

Here is the solution I got working and general approach I'd recommend:
- This is the content of Neovim Lua config (either directly in '~/.config/nvim/init.lua' or in its sourced module):
```lua
-- Set custom descriptions to mappings coming from 'nvim-R'
local has_miniclue, miniclue = pcall(require, 'mini.clue')
if not has_miniclue then return end

miniclue.set_mapping_desc('n', '<localleader>fa', 'Send "echo" "down"')
miniclue.set_mapping_desc('v', '<localleader>fa', 'Send "echo" "down"')
```

- Create '~/.config/nvim/after/ftplugin/r.lua' file with the following content:
```lua
-- Set custom descriptions to mappings coming from 'nvim-R'
local has_miniclue, miniclue = pcall(require, 'mini.clue')
if not has_miniclue then return end

miniclue.set_mapping_desc('n', '<localleader>fa', 'Send "echo" "down"')
miniclue.set_mapping_desc('v', '<localleader>fa', 'Send "echo" "down"')
```

The first step sets up proper 'mini.clue' triggers. The second step is what gets executed when buffer is being set an `r` filetype. Having it in 'after' directory means that this code will be executed *after* all other plugin-specific code. This is the usual approach for user to customize filetype-specific behavior.

Note, that with '~/.config/nvim' this assumes Unix style setup; if you are on Windows, use '~/AppData/Local/nvim'.

------

However, the proper solution here I see is to make an issue/PR to 'Nvim-R' which creates mappings with descriptions. This would require using condition on Neovim>=0.7 (when mapping descriptions were introduced). That repo seems active, so this might result in a success.

**Reply by bassamsdata** [reply]
thank you for your answer, I know it's been a long time but I tried to contribute to NVIM-R regarding their code for the keymaps. I tried multiple solutions regarding the source code but unfortunately didn't succeed at that. tbh, I didn't like vimscript at all.
probably, the simplest solutions, simply :), are the best for many problems.
once again thank you a lot.
